build/scripts/EnlistmentHelperFunctions.psm1

```diff
@@ -315,6 +315,15 @@ function Update-BCArtifactVersion {
 
     Write-Host "Latest BCArtifact URL: $latestArtifactUrl"
 
+    if ($currentArtifactUrl -notlike "https*") {
+        Write-Host "Using latest BCArtifact version for the current repo version: $currentVersion"
+        if ($latestArtifactUrl -contains "bcartifacts") {
+            $latestArtifactUrl = "bcartifacts/Sandbox/$currentVersion/base/latest"
+        } else {
+            $latestArtifactUrl = "bcinsider/Sandbox/$currentVersion/base/latest"
```

Comment:
Shouldn't there be a check to verify this version exists?

E.g. If we bump the version to 24.5, but there is still no 24.5 artifact yet? Should happen around branching, when we change the repo version (and all app versions), but only change the artifact version once there is an artifact with that version.

Reply:
We run `Get-LatestBCArtifactUrl -minimumVersion $currentVersion` earlier in the function. This will fail if there are no 24.5 artifacts.

Reply:
Yes, you are right.

So, this new piece of code just converts a full URL to e.g. `bcinsider/Sandbox/24.5/latest`
Does it then make sense to be part of `Get-LatestBCArtifactUrl`. E.g.
`Get-LatestBCArtifactUrl -minimumVersion 24.4 -asURL` would return the full URL and ``Get-LatestBCArtifactUrl -minimumVersion 24.4` would return... the not full URL 😄 
You got the idea, right?

Also, instead of checking `$currentArtifactUrl -notlike "https*"`, why not check the branch name? That way we won't have to edit the artifact property in 6 release branches and we won't have to remember to change it when branching major versions ((the automation will do all that).

Reply:
> So, this new piece of code just converts a full URL to e.g. bcinsider/Sandbox/24.5/latest
Does it then make sense to be part of Get-LatestBCArtifactUrl. E.g.
Get-LatestBCArtifactUrl -minimumVersion 24.4 -asURL would return the full URL and ``Get-LatestBCArtifactUrl -minimumVersion 24.4` would return... the not full URL 😄
You got the idea, right?

Yeeah I suppose. Except the default behavior of Get-LatestBCArtifact**Url** should probably be to return... a URL 😄 But I can add a `-AsPattern` switch (or something like that) to change this.

> Also, instead of checking $currentArtifactUrl -notlike "https*", why not check the branch name? That way we won't have to edit the artifact property in 6 release branches and we won't have to remember to change it when branching major versions ((the automation will do all that).

My idea was to have the automation respect the whatever we decide the "artifact" setting should be (either URL or pattern). Around the time we branch for a new major version we might want to keep the artifact URLs in the new .x branch for a little bit so we know we're using the right artifacts. There have been cases in the past where we picked up the wrong artifacts. 

As you said, the downside is that I'll have to go update 6 branches once this PR is merged and longer term, we'll have to remember to change it when we branch for a new major version. However, since our automation will try to update the artifact version every day, we will be reminded that we forgot to change it. And once we decide to change it, we can just edit the automated PRs. Shouldn't take long

Reply:
Isn't this exactly why we need the automation to try set the artifact setting based on the branch?

Say we branch 25.x, `artifact` is full URL and then the automation tries to change it to a pattern (the PR will be the reminder).
Same for the cases when we might need to revert to using full URL. The next day the automation will try to set a pattern and the PR build on the automated PR will tell us if it's good to go.

Reply:
> Say we branch 25.x, artifact is full URL and then the automation tries to change it to a pattern (the PR will be the reminder).

That's not ideal if we want to keep getting the full URLs for a few days after we've done branching for 25.x e.g. to make sure we didn't produce an artifact with a higher version out of master before doing the version update to 26.0 in master. Might be nice that the first bcartifact update in 25.x just updates the URL so we know we're picking up a bcartifact produced from 25.x 🤷‍♂️

Reply:
> make sure we didn't produce an artifact with a higher version out of master before doing the version update to 26.0 in master.

That's part of the branching process and shouldn't affect this automation.

I say YOLO, what's the worst that can happen? We built against an older platform and will eventually surface somewhere. Anyway, it is up to you.